rs/node_rewards/canister/src/canister/test/get_node_providers_monthly_xdr_rewards.rs

```diff
@@ -0,0 +1,652 @@
+        &CANISTER,
+        GetNodeProvidersMonthlyXdrRewardsRequest {
+            registry_version: None,
+        },
+    )
+    .now_or_never()
+    .unwrap();
+
+    let monthly_rewards = response.rewards.unwrap();
+    let np1_rewards = monthly_rewards.rewards.get(&np1.0).unwrap();
+    assert_eq!(*np1_rewards, 5); // 5 nodes at 1 XDR/month/node
+    assert_eq!(
+        monthly_rewards.registry_version,
+        Some(registry.latest_version())
+    );
+
+    // Store this version to test specific version requests later.
+    let version_without_rewards_table = registry.latest_version();
+
+    ///////////////////////////////
+    // Now add the reward table for type0 and type2 nodes and check that the values are properly used
+    ///////////////////////////////
+    let json = r#"{
+            "North America,US,NY":         { "type0": [240, null]                                                          },
+            "North America,US":            { "type0": [677, null],  "type2": [456, null]                                   },
+            "North America":               { "type0": [801, null]                                                          },
+            "Europe":                      { "type0": [68, null],   "type2": [11, null]                                    }
+        }"#;
+
+    let map: BTreeMap<String, BTreeMap<String, NodeRewardRate>> =
+        serde_json::from_str(json).unwrap();
+    let node_rewards_payload = UpdateNodeRewardsTableProposalPayload::from(map);
+    update_node_rewards_table(registry.clone(), node_rewards_payload.new_entries);
+
+    let response = NodeRewardsCanister::get_node_providers_monthly_xdr_rewards(
+        &CANISTER,
+        GetNodeProvidersMonthlyXdrRewardsRequest {
+            registry_version: None,
+        },
+    )
+    .now_or_never()
+    .unwrap();
+
+    let monthly_rewards = response.rewards.unwrap();
+
+    // NP1: 4 'type0' nodes in 'North America,US,NY' + 1 'type2' node in 'North America,US'
+    assert_eq!(
+        *monthly_rewards.rewards.get(&np1.0).unwrap(),
+        (4 * 240) + 456
+    );
+    // NP2: 11 'type0' nodes in 'Europe,CH,Zurich' + 7 'type2' nodes in 'CH'
+    assert_eq!(
+        *monthly_rewards.rewards.get(&np2.0).unwrap(),
+        (11 * 68) + (7 * 11)
+    );
+
+    ///////////////////////////////
+    // Now add the reward table entries for type3 rewards and confirm nothing changes for type0 and type2 nodes
+    ///////////////////////////////
+    let json = r#"{
+            "North America,US,NY":         { "type3": [111, null] },
+            "North America,US":            { "type3": [222, null] },
+            "North America":               { "type3": [333, null] },
+            "Europe":                      { "type3": [444, null] },
+            "Asia":                        { "type3": [555, null] }
+        }"#;
+
+    let map: BTreeMap<String, BTreeMap<String, NodeRewardRate>> =
+        serde_json::from_str(json).unwrap();
```

Comment:
Clever. Perhaps I should adopt this json specification style for some Rust tests. Json seems more readable for large tables than Rust.